src/update_engine/omaha_request_action.cc

```diff
@@ -95,26 +95,48 @@ string GetAppBody(const OmahaEvent* event,
   if (event == NULL) {
     app_body = GetPingXml(ping_active_days, ping_roll_call_days);
     if (!ping_only) {
-      app_body += "        <updatecheck></updatecheck>\n";
-
-      // If this is the first update check after a reboot following a previous
-      // update, generate an event containing the previous version number. If
-      // the previous version preference file doesn't exist the event is still
-      // generated with a previous version of 0.0.0.0 -- this is relevant for
-      // older clients or new installs. The previous version event is not sent
-      // for ping-only requests because they come before the client has
-      // rebooted.
+      string current_version;
       string prev_version;
+
+      current_version = params.app_version();
+
       if (!prefs->GetString(kPrefsPreviousVersion, &prev_version)) {
-        prev_version = "0.0.0.0";
+        prev_version = "initial";
+      }
+
+      app_body += "        <updatecheck></updatecheck>\n";
+
+      // Check if currently the same version is running as before the update, which indicates that
+      // a rollback took place.
+      // We do not need to check for UPDATE_STATUS_UPDATED_NEED_REBOOT because a new update won't
+      // be attempted until the system rebooted.
+      if (!prev_version.empty() && prev_version.compare(current_version) == 0) {
```

Comment:
Is `prev_version` ever going to be empty? When we fail to get it from preferences, we set it to `"initial"`.

Reply:
Also, will this branch be executed every time I restart the machine (for whatever reason, not only because of the update) and boot the partition with the previous version? Should we make a check that the update is in progress?

Reply:
`"initial"` should be the value when no file exists while `""` should be the value for normal operation.
Maybe I should introduce another boolean to take care of this? Now I relied on the assumption that `"initial"` won't be the current version.

Reply:
The previous version file is only created when an update was applied. This should not alert a rollback in the case of normal reboots with selecting a different partition manually. (And even if it would fire I think it would be up for discussion if we want to distinguish between automatic and manual rollbacks. Currently I wanted to focus on the first reboot after an update.)

Reply:
> `"initial"` should be the value when no file exists while `""` should be the value for normal operation.
> Maybe I should introduce another boolean to take care of this? Now I relied on the assumption that `"initial"` won't be the current version.

Nah, I'm just not familiar with the code. I was just assuming that this code would also run every time you reboot. And apparently that was a wrong assumption, so let's leave it as is.

> The previous version file is only created when an update was applied. This should not alert a rollback in the case of normal reboots with selecting a different partition manually.

Thanks, I did not know it.

> (And even if it would fire I think it would be up for discussion if we want to distinguish between automatic and manual rollbacks. Currently I wanted to focus on the first reboot after an update.)

Let's have it some other time.